File: app/src/main/kt/estyle/teabaike/activity/SearchActivity.java
package estyle.teabaike.activity;

import android.content.Context;
import android.content.Intent;
import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.support.v4.widget.SwipeRefreshLayout;
import android.view.MenuItem;

import javax.inject.Inject;

import estyle.teabaike.R;
import estyle.teabaike.adapter.MainAdapter;
import estyle.teabaike.application.TeaBaikeApplication;
import estyle.teabaike.databinding.ActivitySearchBinding;
import estyle.teabaike.manager.RetrofitManager;
import estyle.teabaike.widget.FooterView;
import estyle.teabaike.widget.RecyclerView;
import io.reactivex.disposables.Disposable;

public class SearchActivity extends BaseActivity implements MainAdapter.OnItemClickListener,
        SwipeRefreshLayout.OnRefreshListener, RecyclerView.OnLoadMoreListener {

    private ActivitySearchBinding binding;

    private MainAdapter mAdapter;

    private int mPage = 1;
    private String mKeyword;

    @Inject
    RetrofitManager mNetworkProvider;

    private Disposable mDisposable;

    public static void startActivity(Context context, String keyword) {
        Intent intent = new Intent(context, SearchActivity.class);
        intent.putExtra("keyword", keyword);
        context.startActivity(intent);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        TeaBaikeApplication.getInstance().getTeaBaikeComponent().inject(this);
        binding = DataBindingUtil.setContentView(this, R.layout.activity_search);

        initView();
        initData();
    }

    private void initView() {
        setSupportActionBar(binding.toolbar);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);

        binding.swipeRefreshLayout.setColorSchemeResources(R.color.colorAccent);
        binding.swipeRefreshLayout.setOnRefreshListener(this);

        mAdapter = new MainAdapter(this);
        mAdapter.setEmptyView(binding.emptyView);
        mAdapter.setOnItemClickListener(this);

        FooterView footerView = new FooterView(this);
        mAdapter.addFooterView(footerView);

        binding.recyclerView.setOnLoadMoreListener(this);
        binding.recyclerView.setAdapter(mAdapter);
    }

    private void initData() {
        mKeyword = getIntent().getStringExtra("keyword");
        getSupportActionBar().setTitle(mKeyword);

        loadData(mPage, false);
    }

    @Override
    public void onItemClick(int position) {
        ContentActivity.startActivity(this, mAdapter.getItemId(position), true);
    }

    @Override
    public void onRefresh() {
        loadData(mPage = 1, true);
    }

    @Override
    public void onLoadMore() {
        loadData(++mPage, false);
    }

    // 加载网络数据
    private void loadData(int page, final boolean isRefresh) {
        mDisposable = mNetworkProvider.loadSearchData(mKeyword, page)
                .subscribe(dataBeans -> {
                            if (isRefresh) {
                                mAdapter.refreshDatas(dataBeans);
                                binding.swipeRefreshLayout.setRefreshing(false);
                            } else {
                                mAdapter.addDatas(dataBeans);
                                binding.recyclerView.mIsLoading = false;
                            }
                        },
                        throwable -> {
                            binding.emptyView.setText(R.string.fail_connect);
                            if (isRefresh) {
                                binding.swipeRefreshLayout.setRefreshing(false);
                            } else {
                                binding.recyclerView.mIsLoading = false;
                            }
                        });
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                finish();
                break;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        binding.recyclerView.removeOnScrollListener();
        mDisposable.dispose();
    }

}

File: app/src/main/kt/estyle/teabaike/activity/SearchActivity.kt
package estyle.teabaike.activity

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.MenuItem
import androidx.lifecycle.ViewModelProvider
import estyle.base.BaseActivity
import estyle.base.rxjava.DisposableConverter
import estyle.base.rxjava.observer.RefreshObserver
import estyle.base.rxjava.observer.SnackbarObserver
import estyle.base.widget.PagingRecyclerView
import estyle.teabaike.R
import estyle.teabaike.adapter.SearchAdapter
import estyle.teabaike.entity.MainEntity
import estyle.teabaike.viewmodel.SearchViewModel
import kotlinx.android.synthetic.main.activity_search.*

class SearchActivity : BaseActivity(), PagingRecyclerView.OnLoadListener {


    private val viewModel by lazy { ViewModelProvider(this).get(SearchViewModel::class.java) }

    private val adapter by lazy { SearchAdapter() }

    private val keyword by lazy { intent.getStringExtra("keyword") }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_search)

        setSupportActionBar(toolbar)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
        supportActionBar?.title = keyword

        adapter.emptyView = empty_view
        adapter.onItemClickListener = { position, id -> ContentActivity.startActivity(this, id) }
        recycler_view.adapter = adapter

        recycler_view.setOnLoadListener(this)
        swipe_refresh_layout.setOnRefreshListener { refresh() }
    }

    override fun onPostCreate(savedInstanceState: Bundle?) {
        super.onPostCreate(savedInstanceState)

        swipe_refresh_layout.post {
            swipe_refresh_layout.isRefreshing = true
            refresh()
        }
    }

    // PagingRecyclerView
    override fun onLoad() {
        viewModel.loadMore(keyword)
            .`as`(DisposableConverter.dispose(this))
            .subscribe(object : SnackbarObserver<List<MainEntity>>(recycler_view) {

                override fun onNext(it: List<MainEntity>) {
                    super.onNext(it)
                    adapter.load(it)
                    recycler_view.isLoading = false
                }

                override fun onError(e: Throwable) {
                    super.onError(e)
                    recycler_view.isLoading = false
                }
            })
    }

    // SwipeRefreshLayout
    private fun refresh() {
        viewModel.refresh(keyword)
            .`as`(DisposableConverter.dispose(this))
            .subscribe(object :
                RefreshObserver<List<MainEntity>>(swipe_refresh_layout) {
                override fun onNext(it: List<MainEntity>) {
                    super.onNext(it)
                    adapter.refresh(it)
                }

                override fun onError(e: Throwable) {
                    super.onError(e)
                    empty_view.setText(R.string.request_fail)
                }
            })
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> finish()
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onDestroy() {
        super.onDestroy()
        swipe_refresh_layout.setOnRefreshListener(null)
        recycler_view.setOnLoadListener(null)
    }

    companion object {

        fun startActivity(context: Context, keyword: String) {
            val intent = Intent(context, SearchActivity::class.java)
            intent.putExtra("keyword", keyword)
            context.startActivity(intent)
        }
    }
}
